File: acme-java-reactive/src/main/java/acme/repository/AccountRepository.java
package acme.repository;

import acme.entity.Account;
import org.springframework.data.r2dbc.repository.Query;
import org.springframework.data.repository.reactive.ReactiveCrudRepository;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.UUID;

@Repository
public interface AccountRepository extends ReactiveCrudRepository<Account, UUID> {

    @Query("""
        SELECT a.* FROM account a
        JOIN customer_account ca ON a.account_id = ca.account_id
        WHERE ca.customer_id = :customerId
    """)
    Flux<Account> findAccountsByCustomerId(UUID customerId);

    @Query("SELECT SUM(amount) FROM account")
    Mono<Long> getTotalAmount();
}

File: acme-springboot-nonblocking/src/main/kotlin/acme/repository/AccountRepository.kt
package acme.repository

import acme.entity.Account
import kotlinx.coroutines.flow.Flow
import org.springframework.data.r2dbc.repository.Query
import org.springframework.data.repository.kotlin.CoroutineCrudRepository
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.util.*

@Repository
interface AccountRepository : CoroutineCrudRepository<Account, UUID> {

    @Query("""
        SELECT a.* FROM account a
        JOIN customer_account ca ON a.account_id = ca.account_id
        WHERE ca.customer_id = :customerId
    """)
    fun findAccountsByCustomerId(@Param("customerId") customerId: UUID): Flow<Account>

    @Query("SELECT SUM(amount) FROM account")
    suspend fun getTotalAmount(): Long?
}
